lib/stripe/client_api_operations.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module Stripe
+  # Define instance methods on the including class (i.e. StripeClient)
+  # to access API resources.
+  module ClientAPIOperations
+    # Proxy object to inject the client into API resources
+    class ClientProxy
+      def initialize(client:, resource: nil)
+        @client = client
+        @resource = resource
+      end
+
+      attr_reader :client
+
+      def with_client(client)
+        @client = client
+        self
+      end
+
+      def method_missing(method, *args)
+        super unless @resource
+        @resource.send(method, *args << { client: @client }) || super
+      end
+
+      def respond_to_missing?(symbol, include_private = false)
+        super unless @resource
+        @resource.respond_to?(symbol) || super
+      end
+    end
+
+    def self.included(base)
+      base.class_eval do
+        # Sigma, unlike other namespaced API objects, is not separated by a
+        # period so we modify the object name to follow the expected convention.
+        api_resources = Stripe::Util.api_object_classes
+        sigma_class = api_resources.delete("scheduled_query_run")
+        api_resources["sigma.scheduled_query_run"] = sigma_class
+
+        # Group namespaces that have mutiple resourses
+        grouped_resources = api_resources.group_by do |key, _|
+          key.include?(".") ? key.split(".").first : key
+        end
+
+        grouped_resources.each do |resource_namespace, resources|
+          # Namespace resource names are separated with a period by convention.
+          if resources[0][0].include?(".")
+
+            # Defines the methods required for chaining calls for resources that
+            # are namespaced. A proxy object is created so that all resource
+            # methods can be defined at once.
+            proxy = ClientProxy.new(client: nil)
+            resources.each do |resource_name, resource_class|
+              method_name = resource_name.split(".").last
+              proxy.define_singleton_method("#{method_name}s") do
```

Comment:
This is a little bike shed, but I kinda think we might want to drop the pluralization just because (1) doesn't make that much of a difference, (2) cancels the possibility of needing a more complex pluralization engine for nouns with more complex pluralization rules, and (3) introduces the invariant of _resource name == client proxy name_ for maximum obviousness.

I believe (2) might already be a problem in for example "capability" which probably lands as "capabilitys" instead of "capabilities".

cc @ob-stripe @remi-stripe In case they have opinions on this one though because we do pluralize in some languages already like stripe-node (e.g. `stripe.charges.list(...)`).

Reply:
And hmm, okay, I'd still stand by pluralization in general being kind of a bad idea, but it turns out that in at least a few other languages (PHP, Node) you access via pluralized resource names, so we should leave it pluralized here as well.

Reply:
Mostly opted for pluralization due to seeing it in a few other Stripe clients that have this sort of API. Sounds like we'll leave it?

I had a feeling the pluralization logic was too naive and completely forgot spaced on pluralization being more complex than adding an `s`. Would you prefer a smarter pluralization util over adding pluralized names to each resource?

Reply:
Yeah, ignore what I said originally :) Let's leave it pluralized for consistency.

And nah — don't worry about adding a util for now, although maybe put a comment indicating that at some point a smarter pluralization scheme might be necessary. It turns out that capabilities are always accessed as a subresource like `Stripe::Account.list_capabilities`, so it's not a problem for them, and I went through the rest of our resources and I don't think there are any other (although some come close, liked "scheduled queries", but luckily we have "run" at the end; "person" is accessed as "persons" instead of "people").

Flagging to @ob-stripe and @remi-stripe again though that there is some possibility that this introduces a problem in the future if we're not careful because if we do add a resource with a non-standard pluralization scheme and forget to update this, it'll become automatically accessible at the "wrong" pluralization. If/when we realize that later, fixing it would be backwards incompatible.